native/src/main/java/io/ballerina/stdlib/data/csvdata/csv/CsvTraversal.java

```diff
@@ -0,0 +1,802 @@
+                BArray array = (BArray) csvElement;
+                constructArrayValuesFromArray(array, type, expectedTypeSize == -1 ? array.size() : expectedTypeSize);
+            }
+        }
+
+        private void constructArrayValuesFromArray(BArray csvElement, Type type, int expectedSize) {
+            int index = 0;
+            for (int i = 0; i < csvElement.getLength(); i++) {
+                if (config.allowDataProjection && index >= expectedSize) {
+                    break;
+                }
+                Type memberType = getArrayOrTupleMemberType(type, index);
+                if (memberType != null) {
+                    addValuesToArrayType(csvElement.get(i), memberType, index,
+                            currentCsvNode, config);
+                }
+                index++;
+            }
+        }
+
+        private void constructArrayValuesFromMap(BMap<BString, Object> map, Type type, int expectedSize) {
+            int size = map.size();
+            BString[] keys = new BString[size];
+            int index = 0;
+            if (config.headersOrder != null) {
+                String[] headerOrder = config.headersOrder.getStringArray();
+                if (headerOrder.length != size) {
+                    throw DiagnosticLog.error(DiagnosticErrorCode.INVALID_HEADER_NAMES_LENGTH);
+                }
+                for (int i = 0; i < size; i++) {
+                    keys[i] = StringUtils.fromString(headerOrder[i]);
+                }
+            } else if (config.customHeader == null) {
+                keys = map.getKeys();
+            } else {
+                if (this.headers == null) {
+                    this.headers = createHeaders(new String[size], config);
+                }
+                if (this.headers.length != size) {
+                    throw DiagnosticLog.error(DiagnosticErrorCode.INVALID_CUSTOM_HEADER_LENGTH);
+                }
+                for (int i = 0; i < size; i++) {
+                    keys[i] = StringUtils.fromString(this.headers[i]);
+                }
+            }
+            for (BString key: keys) {
+                if (!map.containsKey(key)) {
+                    throw DiagnosticLog.error(DiagnosticErrorCode.INVALID_CUSTOM_HEADER, key);
+                }
+                Object v = map.get(key);
+                if (config.allowDataProjection && index >= expectedSize) {
+                    break;
+                }
+                Type memberType = getArrayOrTupleMemberType(type, index);
+                if (memberType != null) {
+                    addValuesToArrayType(v, memberType, index,
+                            currentCsvNode, config);
+                }
+                index++;
+            }
+        }
+
+        private boolean isArrayOrTupleRestTypeMember(Type type, int index) {
+            if (type instanceof ArrayType) {
+                return false;
+            }
+            return ((TupleType) type).getTupleTypes().size() < index + 1;
```

Comment:
Minor nit, I think you can invert the condition to avoid a extra type check
```suggestion
            if (type instance of TupleType tupleType) {
               tupleType.getTupleTypes().size() < index + 1;
            }
            return false;
```